src/cache.rs

```diff
@@ -0,0 +1,46 @@
+use crate::errors::*;
+use crate::metrics::{Counter, MetricOpts, Metrics};
+use bitcoin_hashes::sha256d::Hash as Sha256dHash;
+use lru::LruCache;
+use std::sync::Mutex;
+
+pub struct BlockTxIDsCache {
+    map: Mutex<LruCache<Sha256dHash /* blockhash */, Vec<Sha256dHash /* txid */>>>,
+    hits: Counter,
+    misses: Counter,
+}
+
+impl BlockTxIDsCache {
+    pub fn new(capacity: usize, metrics: &Metrics) -> BlockTxIDsCache {
+        BlockTxIDsCache {
+            map: Mutex::new(LruCache::new(capacity)),
+            hits: metrics.counter(MetricOpts::new(
+                "electrs_blocktxids_cache_hits",
+                "# of cache hits for list of transactions in a block",
+            )),
+            misses: metrics.counter(MetricOpts::new(
+                "electrs_blocktxids_cache_misses",
+                "# of cache misses for list of transactions in a block",
+            )),
+        }
+    }
+
+    pub fn get_or_else<F>(
+        &self,
+        blockhash: &Sha256dHash,
+        load_txids_func: F,
+    ) -> Result<Vec<Sha256dHash>>
+    where
+        F: FnOnce() -> Result<Vec<Sha256dHash>>,
+    {
+        if let Some(txids) = self.map.lock().unwrap().get(blockhash) {
+            self.hits.inc();
+            return Ok(txids.clone());
```

Comment:
Is it possible to avoid the `.clone()` here, by returning a `Result<&[Sha256dHash]>`?

Reply:
I'm not sure how. `LruCache::get` returns a reference.

Reply:
Maybe a way to do this would be to instead of returning the transaction list for processing, pass a callback for processing the transaction list in place, for example:
```
let result = self.app.daemon().withblocktxids(&header_entry.hash(), |txids| {
    let pos = txids
       .iter()
       .position(|txid| txid == tx_hash)
       .chain_err(|| format!("missing txid {}", tx_hash))?;
   let (branch, _root) = create_merkle_branch_and_root(txids, pos);
   Ok((branch, pos))
}
```
The trade-off is keeping a lock on the cache for a longer time. Also probably better to do in a separate PR.

Reply:
> Maybe a way to do this would be to instead of returning the transaction list for processing, pass a callback for processing the transaction list in place

Good solution!

> Also probably better to do in a separate PR.

Agreed :)